Added implementation for copy, cut and paste

RobertGrancsa opened the pull request:
Using the usual ctrl+c, ctrl+x, ctrl+v, you can copy, cut and paste the new lines. I implemented a character buffer that copies the line into every time you press copy or cut, and can paste it anywhere in the file.
Reused this implementation of dublicate, and now duplicate is just a combination of the copy and paste methods, so as to reduce the number of duplicate lines.

Signed-off-by: Robert Grancsa <robert.grancsa2002@gmail.com>

alexandruradovici commented:
Looks good, does it fit now into the 1024 lines?

alexandruradovici commented on src/editor.rs:
```diff
@@ -408,18 +413,34 @@ impl Editor {
     }
 
+	fn cut_current_row(&mut self) {
+        self.copy_current_row();
+		self.delete_current_row();
```
Run `cargo fmt` to update format.

alexandruradovici commented on src/editor.rs:
```diff
@@ -408,18 +413,34 @@ impl Editor {
     fn duplicate_current_row(&mut self) {
+        self.copy_current_row();
+		self.paste_current_row();
+		self.copied_row.clear();
```
If the source is longer than 1024 lines, this line may be deleted.

RobertGrancsa commented:
I reformatted the lines, and deleted the suggested line, but the line count is still a bit over 1024 (1028 to be exact, but that only for unix systems, windows and wasi are under that line number)

alexandruradovici commented:
How many lines are in the original code?

RobertGrancsa commented:
1011 lines in UNIX

alexandruradovici commented:
> 1011 lines in UNIX

So you added an extra 17 lines?

RobertGrancsa commented on src/editor.rs:
```diff
@@ -408,16 +413,29 @@ impl Editor {
+
+    fn paste_current_row(&mut self) {
+        if self.copied_row.is_empty() { return; }
+        let new_row = Row::new(self.copied_row.clone());
```
As in I shouldn't check if it's empty or not? Because if I don't do that check, it will insert a new empty line each time the copy buffer is empty

alexandruradovici replied:
You should, but then use it directly.

RobertGrancsa commented on src/editor.rs:
```diff
@@ -408,16 +413,29 @@ impl Editor {
+        self.paste_current_row();
+    }
+
+    fn copy_current_row(&mut self) { 
+        if let Some(row) = self.current_row() { self.copied_row = row.chars.clone(); }
+    }
+
+    fn cut_current_row(&mut self) {
+        self.copy_current_row();
+        self.delete_current_row();
+    }
```
It just lowered the number of lines by one single line, now I am at 1027

alexandruradovici commented on src/editor.rs:
```diff
@@ -408,16 +413,29 @@ impl Editor {
+        self.paste_current_row();
+    }
+
+    fn copy_current_row(&mut self) { 
+        if let Some(row) = self.current_row() { self.copied_row = row.chars.clone(); }
+    }
+
+    fn cut_current_row(&mut self) {
+        self.copy_current_row();
+        self.delete_current_row();
+    }
```
```suggestion
    fn copy_or_cut_current_row(&mut self, copy: bool) { 
        if let Some(row) = self.current_row() { self.copied_row = row.chars.clone(); }
        if (!copy) { self.delete_current_row(); }
    }
```

RobertGrancsa commented:
I think everything should be good now, everything is under 1024 lines and the commands work as intended

alexandruradovici commented on src/editor.rs:
```diff
@@ -15,13 +15,16 @@ const REFRESH_SCREEN: u8 = ctrl_key(b'L');
 
 const HELP_MESSAGE: &str =
-    "Ctrl-S = save | Ctrl-Q = quit | Ctrl-F = find | Ctrl-G = go to | Ctrl-D = duplicate | Ctrl-E = execute";
+    "Ctrl-S = save | Ctrl-Q = quit | Ctrl-F = find | Ctrl-G = go to | Ctrl-D = duplicate | Ctrl-E = execute | Ctrl-C = copy | Ctrl-X = cut | Ctrl-V = paste";
```
Does this line fit into the screen? What about using `^S = save | ^Q = quit ...`

RobertGrancsa replied:
Alright, I'm changing it right now

alexandruradovici commented:
This looks like it is ready. Please modify PR text describing what all the modifications you have made. As soon as we have the text, I suggest opening a PR to upstream.

alexandruradovici reviewed (changes requested):
Formatting seems to be wrong, please run

```
rustup install nightly
cargo +nightly fmt
```

RobertGrancsa commented:
By running the `cargo +nightly fmt,` some braces will open up, and the number of lines become longer than 1024. What should I do about this?

alexandruradovici commented:
Yes but otherwise your checks are failing due to the format. How many extra lines do you have?

RobertGrancsa commented:
1030 after the formatting, 1022 before

alexandruradovici commented:
If you look at the original upstream code, how many lines do you get?

RobertGrancsa commented:
1011 lines in the original one

RobertGrancsa commented:
Found a way to skip the format checking of the function with `#[rustfmt::skip]`, if that works and is in no way disrupting something in that case, and the tests should run this way

alexandruradovici commented:
> Found a way to skip the format checking of the function with `#[rustfmt::skip]`, if that works and is in no way disrupting something in that case, and the tests should run this way

I don't think the upstream will accept this. Can you play around with rustfmt.toml, maybe we can add or tweak some rule?

RobertGrancsa commented:
Can't seem to possibly find any other way of compressing the code, it already has all the flags that make it the smallest it can be

alexandruradovici commented:
You can group assignments together like

https://github.com/UPB-CS-OpenSourceUpstream/kibi/blob/2cc6bead72e2b48203d6ddf28f189c51868a9a79/src/unix.rs#L74-L77

```rust
(term.c_cc[VMIN], term.c_cc[VTIME]) = (0, 1);
```

RobertGrancsa commented:
I think it should be all good now, everything is under 1024 lines

alexandruradovici commented on src/editor.rs:
```diff
@@ -286,8 +285,7 @@ impl Editor {
         self.screen_rows = wsize.0.saturating_sub(2); // Make room for the status bar and status message
-        self.window_width = wsize.1;
-        self.update_screen_cols();
+        (self.window_width, _) = (wsize.1, self.update_screen_cols());
```
This is unstable, fails build.

alexandruradovici commented on src/editor.rs:
```diff
@@ -395,41 +392,45 @@ impl Editor {
             self.update_row(self.cursor.y, false);
-            self.cursor.move_to_next_line();
-            self.delete_char();
+            (_, _) = (self.cursor.move_to_next_line(), self.delete_char());
```
This is unstable, doesn't build.

alexandruradovici commented on src/editor.rs:
```diff
@@ -395,41 +392,45 @@ impl Editor {
     fn duplicate_current_row(&mut self) {
-        self.copy_current_row(true);
-        self.paste_current_row();
+        (_, _) = (self.copy_current_row(true), self.paste_current_row());
```
This is unstable, doesn't build.

alexandruradovici commented on src/editor.rs:
```diff
@@ -395,41 +392,45 @@ impl Editor {
-        self.dirty = true;
         // The line number has changed
-        self.update_screen_cols();
+        (self.cursor.y, self.dirty, _) = (self.cursor.y + 1, true, self.update_screen_cols());
```
Use only for the assignments, as the rest is unstable.

RobertGrancsa replied:
Got it

alexandruradovici commented:
Please run the format tool before you submit, you still have formatting issues.

RobertGrancsa commented:
Sorry, now it should all be good

alexandruradovici commented:
This change requires rust version 1.95, https://blog.rust-lang.org/2022/02/24/Rust-1.59.0.html. Please change the ci.yml file and Readme file to reflect that

RobertGrancsa commented:
Does it work like this?

alexandruradovici commented:
Please fix the linter's issues.

RobertGrancsa commented:
This will be my PR text:

**Added implementation for copy, cut and paste**

Using the usual ctrl+c, ctrl+x, ctrl+v, you can copy, cut and paste the new lines. I implemented a character buffer that copies the line into every time you press copy or cut, and can paste it anywhere in the file.
Reused the implementation of duplicate, and now duplicate is just a combination of the copy and paste methods, so as to reduce the number of duplicate lines.

Updated the README.md to reflect these changes. Also had to update the minimum rust version from 1.53 to 1.59, because of some unstable behavior found when double assigning variables, so as to reduce the number of lines. 

Some other clippy issues were found, regarding some borrowing and some casting, which now should be solved

Signed-off-by: Robert Grancsa [robert.grancsa2002@gmail.com](mailto:robert.grancsa2002@gmail.com)

alexandruradovici reviewed (approved):
This should be ready to be pushed. Open a pull request to the original repository and write a link in this pull request discussion.

RobertGrancsa commented:
Thank you for all the help

Here is the link for the PR: https://github.com/ilai-deutel/kibi/pull/207

alexandruradovici commented:
1. What's with the merge?
2. Please edit your PR text so that the issue that you fix is connected (# 22 without the space) and delete my comment there.

RobertGrancsa commented:
There were some updates for some files in the master branch, and they updated the minimum version of rust to 1.61, and I had to change some files because there were some merging errors.

alexandruradovici commented:
This statement "I updated the minimum rust version from 1.53 to 1.59, because I needed the destructuring assignment (https://blog.rust-lang.org/2022/02/24/Rust-1.59.0.html), so as to reduce the number of lines." is not valid anymore, as they have updated rust in the upstream repository.

RobertGrancsa commented:
They approved the changes and merged the branch. Thank you again for all the help, it was a pleasure working on this feature and I hope I have the time to also try some other problems.

alexandruradovici commented:
Closing as https://github.com/ilai-deutel/kibi/pull/207

alexandruradovici commented:
Congratulations, what about trying another issue?

alexandruradovici commented:
@RobertGrancsa Please select a timeslot for the interview https://doodle.com/meeting/participate/id/eZ8gOOEb.
